app/models/post_redirect.rb

```diff
@@ -80,6 +89,22 @@ def local_part_uri
     $1
   end
 
+  def email_token_valid?
+    return true if circumstance == 'normal'
+
+    case circumstance
+    when 'change_email'
+      value = user.email
+    when 'change_password'
+      value = user.hashed_password
+    end
+
+    data = PostRedirect.verifier.verify(email_token)
+    nonce = Digest::MD5.hexdigest(value)
+
+    user.id == data[:user_id] && nonce == data[:nonce]
```

Comment:
Do we need to compare the `user.id` or even store in the token?

Reply:
I'm going to keep this in. In case the digest value is identical for different users (shouldn't happen for the hashed password but could for the email address I guess or a MD5 collision)